ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts

```diff
@@ -0,0 +1,145 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { GridItemDefinition, LayoutDefinition } from '@perses-dev/core';
+import { StateCreator } from 'zustand';
+import { Middleware } from './common';
+
+/**
+ * Slice with the state of Panel Groups, as well as any actions that modify only Panel Group state.
+ */
+export interface PanelGroupSlice {
+  /**
+   * Panel groups indexed by their ID.
+   */
+  panelGroups: Record<PanelGroupId, PanelGroupDefinition>;
+
+  /**
+   * An array of panel group IDs, representing their order in the dashboard.
+   */
+  panelGroupIdOrder: PanelGroupId[];
+
+  /**
+   * previous state
+   */
+  previousPanelGroupStates: {
+    panelGroups: PanelGroupSlice['panelGroups'];
+    panelGroupIdOrder: PanelGroupSlice['panelGroupIdOrder'];
+  };
+
+  // TODO: Remove this
+  createPanelGroupId: () => PanelGroupId;
+
+  /**
+   * Rearrange the order of panel groups by swapping the positions
+   */
+  swapPanelGroups: (xIndex: number, yIndex: number) => void;
+
+  /**
+   * save
+   */
+  savePanelGroups: () => void;
+
+  /**
+   * reset to previous panel group states
+   */
+  resetPanelGroups: () => void;
+}
+
+export type PanelGroupId = number;
+
+export interface PanelGroupDefinition {
+  id: PanelGroupId;
+  items: GridItemDefinition[];
+  isCollapsed: boolean;
+  title?: string;
+}
+
+/**
+ * Uniquely identifies an item in a PanelGroup.
+ */
+export interface PanelGroupItemId {
+  panelGroupId: PanelGroupId;
+  itemIndex: number;
+}
```

Comment:
OK my idea here would be:

`export interface PanelGroupItem` (because this encapsulates more than an ID)
`panelGroupId: PanelGroupID`
`index: number` (since we're inside PanelGroupItem, `index` seems enough)

Reply:
Actually I take it back `itemIndex` seems nice too

Reply:
I'm just spitballing now, but I wonder if we'd want to even have:

`panelGroupId: PanelGroupID`
`panelIndex: number`

Have not thought this all the way through, but maybe we want to stop saying generic "item" and just starting saying "panel"

Reply:
Yeah, I'm thinking when I split out "items" so that I can connect drag and drop, there may be some renaming here. Look for that in a PR soon.